python/ray/ml/checkpoint.py

```diff
@@ -1,26 +1,238 @@
+            mlflow.sklearn.save_model(clf, "model_directory")
+
+            # Create checkpoint object from path
+            checkpoint = Checkpoint.from_directory("model_directory")
+
+            # Convert into dictionary
+            checkpoint_dict = checkpoint.to_dict()
+
+            # This dict can then be passed around, e.g. to a different function
+
+            # At some other location, recover checkpoint object from dict
+            checkpoint = Checkpoint.from_dict(checkpoint_dict)
+
+            # Convert into a directory again
+            checkpoint.to_directory("other_directory")
+
+            # We can now use MLflow to re-load the model
+            clf = mlflow.sklearn.load_model("other_directory")
+
+            # It is guaranteed that the original data was recovered
+            assert isinstance(clf, RandomForestClassifier)
+
+
+    """
+
+    @DeveloperAPI
+    def __init__(
+        self,
+        local_path: Optional[str] = None,
+        data_dict: Optional[dict] = None,
+        uri: Optional[str] = None,
+        obj_ref: Optional[ray.ObjectRef] = None,
+    ):
+        # First, resolve file:// URIs to local paths
+        if uri:
+            local_path = _get_local_path(uri)
+            if local_path:
+                uri = None
+
+        # Only one data type can be set at any time
+        if local_path:
+            assert not data_dict and not uri and not obj_ref
+            if not isinstance(local_path, (str, os.PathLike)) or not os.path.exists(
+                local_path
+            ):
+                raise RuntimeError(
+                    f"Cannot create checkpoint from path as it does "
+                    f"not exist on local node: {local_path}"
+                )
+        elif data_dict:
+            assert not local_path and not uri and not obj_ref
+            if not isinstance(data_dict, dict):
+                raise RuntimeError(
+                    f"Cannot create checkpoint from dict as no "
+                    f"dict was passed: {data_dict}"
+                )
+        elif obj_ref:
+            assert not local_path and not data_dict and not uri
+            if not isinstance(obj_ref, ray.ObjectRef):
+                raise RuntimeError(
+                    f"Cannot create checkpoint from object ref as no "
+                    f"object ref was passed: {obj_ref}"
+                )
+        elif uri:
+            assert not local_path and not data_dict and not obj_ref
+            uri = _get_external_path(uri)
+            if not uri:
+                raise RuntimeError(
+                    f"Cannot create checkpoint from URI as it is not "
+                    f"supported: {uri}"
+                )
+        else:
+            raise ValueError("Cannot create checkpoint without data.")
+
+        self._local_path = local_path
+        self._data_dict = data_dict
+        self._uri = uri
+        self._obj_ref = obj_ref
+
+    def get_local_path(self) -> Optional[str]:
```

Comment:
Instead of adding get_* for each representation, how about adding a common `get_internal_representation()` call that returns a tuple of the representation format and the value, e.g., `("uri", "s3://foo/bar")`?

We can mark this as DeveloperAPI and note that it is only for advanced usage. This avoids confusion with the more common public methods of Checkpoint that most users will want to know about.

Also can move to the bottom of the class.

Reply:
S.g., updated